### ISSUE TITLE
Variable value (label) of OP(fluid.layers.cross_entropy) expected >= 0 and < 1, but got 1.

### ISSUE THREAD

**Issue by FwtGit** [PaddlePaddle/Paddle#23003]
在使用交叉熵损失函数的时候出现这个错误，交叉熵label的范围是否应为 >= 0 and < =1？
因为我在做一个二分类的任务，label不是0就是1，想要用这个损失函数，不知道应该怎么用，烦请解答！谢谢！

**Comment by LDOUBLEV** [comment]
你是不是设置了soft_label，方便提供下报错代码吗

**Comment by FwtGit** [comment]
> 你是不是设置了soft_label，方便提供下报错代码吗

代码如下

import paddle
import paddle.fluid as fluid
import numpy as np
import random



def train_reader():
    def _reader_():
        f= open("data/data23043/train1w.txt")
        line =f.readline()
        userdic={}
        train_ps=[]
        train_ng=[]
        ##读取正采样
        print("载入训练数据.....")
        while line:
            user,item,label = line.strip("\n").split("\t")
            train_ps.append([user,item,int(label)])
            if user in userdic.keys():
                userdic[user].append(item)
            else:
                userdic[user]=[]
                userdic[user].append(item)
            line= f.readline()
        print("负采样......")
        ##负采样                                ##注意 songlist 里没有 6582,7359
        for user in userdic.keys():
            for _  in range(4):
                ng = random.randint(1,6581)
                while str(ng) in userdic[user]  or ng==6582 or ng ==7359:
                    ng = random.randint(1, 6581)
                train_ng.append([user,ng,0])
        train=train_ps+train_ng
        for idx in range(len(train)):
            yield train[idx]
    return  _reader_

def test_reader():
    def reader():
        f = open("data/data23043/test1w.txt")
        line = f.readline()
        test=[]
        while line:
            user, item, label = line.strip("\n").split("\t")
            test.append([user,item,int(label)])
            line =f.readline()
        for idx in range(len(test)):
            yield test[idx]
    return  reader



class NCF(fluid.dygraph.Layer):
    def __init__(self,user_num,item_num,factor_num,num_layers,dropout):
        super(NCF, self).__init__()

        self.embed_user_GMF = fluid.dygraph.Embedding(size=[user_num,factor_num])
        self.embed_item_GMF = fluid.dygraph.Embedding(size=[item_num,factor_num])
        self.embed_user_MLP = fluid.dygraph.Embedding(size=[user_num, factor_num * (2 ** (num_layers - 1))])
        self.embed_item_MLP = fluid.dygraph.Embedding(size=[item_num, factor_num * (2 ** (num_layers - 1))])
        MLP_modules=[]
        for i in range(num_layers):
            input_size = factor_num * (2 ** (num_layers - i))
            # MLP_modules.append(fluid.layers.dropout(dropout_prob=dropout))
            MLP_modules.append(fluid.dygraph.Linear(input_dim=input_size,output_dim=input_size//2,act="relu"))
        self.MLP_layers = fluid.dygraph.Sequential(*MLP_modules)
        self.predict_layer = fluid.dygraph.Linear(input_dim=factor_num*2,output_dim=1)


    def forward(self, user,item):
        embed_user_GMF = self.embed_user_GMF(user)
        embed_item_GMF = self.embed_item_GMF(item)
        output_GMF = embed_user_GMF * embed_item_GMF

        embed_user_MLP = self.embed_user_MLP(user)
        embed_item_MLP = self.embed_item_MLP(item)
        interaction =  fluid.layers.concat(input=[embed_user_MLP, embed_item_MLP],axis=-1)
        output_MLP = self.MLP_layers(interaction)

        concat = fluid.layers.concat(input=[output_GMF, output_MLP], axis=-1)
        prediction = self.predict_layer(concat)
        return prediction


if __name__ == '__main__':
    ##参数初始化
    BATCH_SIZE = 16384
    user_num = 30754
    item_num = 10001
    factor_num = 32
    num_layers = 2
    dropout = 0.0
    epoch_num = 100
    ##训练数据准备
    train_data = train_reader()
    train_suffle = paddle.reader.shuffle(train_data,buf_size=BATCH_SIZE*2)
    train_reader = paddle.batch(train_suffle,batch_size=BATCH_SIZE)
    ##准备测试数据
    test_Data = test_reader()
    test_reader = paddle.batch(test_Data,batch_size=100)

with fluid.dygraph.guard(fluid.CPUPlace()):
    ##定义模型
    model = NCF(user_num=user_num, item_num=item_num,
                factor_num=factor_num, num_layers=num_layers, dropout=dropout)
    ##定义优化器
    adam = fluid.optimizer.AdamOptimizer(parameter_list=model.parameters())
   
   
    for epoch in range (epoch_num):
        model.train()
        for batch_id,data in enumerate(train_reader()):
            data =np.asarray(data).T
            user = fluid.dygraph.to_variable(data[0].reshape([-1,1]).astype("int64"))
            item = fluid.dygraph.to_variable(data[1].reshape([-1,1]).astype("int64"))
            label = fluid.dygraph.to_variable(data[2].reshape([-1,1]).astype("int64"))
            prediction = model(user,item)
            loss = fluid.layers.cross_entropy(input=prediction,label=label)
            # loss = fluid.layers.mse_loss(input=prediction,label=label)
            # loss = fluid.layers.square_error_cost(input=prediction,label=label)
            avg_loss = fluid.layers.mean(loss)
            # print(avg_loss.numpy())
            loss.backward()
            adam.minimize(avg_loss)
            model.clear_gradients()

        model.eval()
        hit = 0
        num = 0
        for batch_id, data in enumerate(test_reader()):
            data = np.asarray(data).T
            user = fluid.dygraph.to_variable(data[0].reshape([-1, 1]).astype("int64"))
            item = fluid.dygraph.to_variable(data[1].reshape([-1, 1]).astype("int64"))
            label = fluid.dygraph.to_variable(data[2].reshape([-1, 1]).astype("int64"))
            prediction = model(user, item)
            pre = fluid.layers.reshape(x=prediction,shape=[1,-1])
            _, topk = fluid.layers.topk(input=pre, k=10)
            topk =list(topk.numpy())[0]
            if 0 in topk:
                hit = hit + 1
                num = num + 1
            else:
                num = num + 1
            # if (batch_id % 10000 == 0):
            #     print(topk)
        print("epoch:%d,HR@10=%f" % (epoch, hit / num))

**Comment by FwtGit** [comment]
> 你是不是设置了soft_label，方便提供下报错代码吗

`import paddle
import paddle.fluid as fluid
import numpy as np
import random



def train_reader():
    def _reader_():
        f= open("data/data23043/train1w.txt")
        line =f.readline()
        userdic={}
        train_ps=[]
        train_ng=[]
        ##读取正采样
        print("载入训练数据.....")
        while line:
            user,item,label = line.strip("\n").split("\t")
            train_ps.append([user,item,int(label)])
            if user in userdic.keys():
                userdic[user].append(item)
            else:
                userdic[user]=[]
                userdic[user].append(item)
            line= f.readline()
        print("负采样......")
        ##负采样                                ##注意 songlist 里没有 6582,7359
        for user in userdic.keys():
            for _  in range(4):
                ng = random.randint(1,6581)
                while str(ng) in userdic[user]  or ng==6582 or ng ==7359:
                    ng = random.randint(1, 6581)
                train_ng.append([user,ng,0])
        train=train_ps+train_ng
        for idx in range(len(train)):
            yield train[idx]
    return  _reader_

def test_reader():
    def reader():
        f = open("data/data23043/test1w.txt")
        line = f.readline()
        test=[]
        while line:
            user, item, label = line.strip("\n").split("\t")
            test.append([user,item,int(label)])
            line =f.readline()
        for idx in range(len(test)):
            yield test[idx]
    return  reader



class NCF(fluid.dygraph.Layer):
    def __init__(self,user_num,item_num,factor_num,num_layers,dropout):
        super(NCF, self).__init__()

        self.embed_user_GMF = fluid.dygraph.Embedding(size=[user_num,factor_num])
        self.embed_item_GMF = fluid.dygraph.Embedding(size=[item_num,factor_num])
        self.embed_user_MLP = fluid.dygraph.Embedding(size=[user_num, factor_num * (2 ** (num_layers - 1))])
        self.embed_item_MLP = fluid.dygraph.Embedding(size=[item_num, factor_num * (2 ** (num_layers - 1))])
        MLP_modules=[]
        for i in range(num_layers):
            input_size = factor_num * (2 ** (num_layers - i))
            # MLP_modules.append(fluid.layers.dropout(dropout_prob=dropout))
            MLP_modules.append(fluid.dygraph.Linear(input_dim=input_size,output_dim=input_size//2,act="relu"))
        self.MLP_layers = fluid.dygraph.Sequential(*MLP_modules)
        self.predict_layer = fluid.dygraph.Linear(input_dim=factor_num*2,output_dim=1)


    def forward(self, user,item):
        embed_user_GMF = self.embed_user_GMF(user)
        embed_item_GMF = self.embed_item_GMF(item)
        output_GMF = embed_user_GMF * embed_item_GMF

        embed_user_MLP = self.embed_user_MLP(user)
        embed_item_MLP = self.embed_item_MLP(item)
        interaction =  fluid.layers.concat(input=[embed_user_MLP, embed_item_MLP],axis=-1)
        output_MLP = self.MLP_layers(interaction)

        concat = fluid.layers.concat(input=[output_GMF, output_MLP], axis=-1)
        prediction = self.predict_layer(concat)
        return prediction


if __name__ == '__main__':
    ##参数初始化
    BATCH_SIZE = 16384
    user_num = 30754
    item_num = 10001
    factor_num = 32
    num_layers = 2
    dropout = 0.0
    epoch_num = 100
    ##训练数据准备
    train_data = train_reader()
    train_suffle = paddle.reader.shuffle(train_data,buf_size=BATCH_SIZE*2)
    train_reader = paddle.batch(train_suffle,batch_size=BATCH_SIZE)
    ##准备测试数据
    test_Data = test_reader()
    test_reader = paddle.batch(test_Data,batch_size=100)

with fluid.dygraph.guard(fluid.CPUPlace()):
    ##定义模型
    model = NCF(user_num=user_num, item_num=item_num,
                factor_num=factor_num, num_layers=num_layers, dropout=dropout)
    ##定义优化器
    adam = fluid.optimizer.AdamOptimizer(parameter_list=model.parameters())
   
   
    for epoch in range (epoch_num):
        model.train()
        for batch_id,data in enumerate(train_reader()):
            data =np.asarray(data).T
            user = fluid.dygraph.to_variable(data[0].reshape([-1,1]).astype("int64"))
            item = fluid.dygraph.to_variable(data[1].reshape([-1,1]).astype("int64"))
            label = fluid.dygraph.to_variable(data[2].reshape([-1,1]).astype("int64"))
            prediction = model(user,item)
            loss = fluid.layers.cross_entropy(input=prediction,label=label)
            # loss = fluid.layers.mse_loss(input=prediction,label=label)
            # loss = fluid.layers.square_error_cost(input=prediction,label=label)
            avg_loss = fluid.layers.mean(loss)
            # print(avg_loss.numpy())
            loss.backward()
            adam.minimize(avg_loss)
            model.clear_gradients()

        model.eval()
        hit = 0
        num = 0
        for batch_id, data in enumerate(test_reader()):
            data = np.asarray(data).T
            user = fluid.dygraph.to_variable(data[0].reshape([-1, 1]).astype("int64"))
            item = fluid.dygraph.to_variable(data[1].reshape([-1, 1]).astype("int64"))
            label = fluid.dygraph.to_variable(data[2].reshape([-1, 1]).astype("int64"))
            prediction = model(user, item)
            pre = fluid.layers.reshape(x=prediction,shape=[1,-1])
            _, topk = fluid.layers.topk(input=pre, k=10)
            topk =list(topk.numpy())[0]
            if 0 in topk:
                hit = hit + 1
                num = num + 1
            else:
                num = num + 1
            # if (batch_id % 10000 == 0):
            #     print(topk)
        print("epoch:%d,HR@10=%f" % (epoch, hit / num))
`

**Comment by FwtGit** [comment]
模型定义：


class NCF(fluid.dygraph.Layer):
    def __init__(self,user_num,item_num,factor_num,num_layers,dropout):
        super(NCF, self).__init__()

        self.embed_user_GMF = fluid.dygraph.Embedding(size=[user_num,factor_num])
        self.embed_item_GMF = fluid.dygraph.Embedding(size=[item_num,factor_num])
        self.embed_user_MLP = fluid.dygraph.Embedding(size=[user_num, factor_num * (2 ** (num_layers - 1))])
        self.embed_item_MLP = fluid.dygraph.Embedding(size=[item_num, factor_num * (2 ** (num_layers - 1))])
        MLP_modules=[]
        for i in range(num_layers):
            input_size = factor_num * (2 ** (num_layers - i))
            # MLP_modules.append(fluid.layers.dropout(dropout_prob=dropout))
            MLP_modules.append(fluid.dygraph.Linear(input_dim=input_size,output_dim=input_size//2,act="relu"))
        self.MLP_layers = fluid.dygraph.Sequential(*MLP_modules)
        self.predict_layer = fluid.dygraph.Linear(input_dim=factor_num*2,output_dim=1)


    def forward(self, user,item):
        embed_user_GMF = self.embed_user_GMF(user)
        embed_item_GMF = self.embed_item_GMF(item)
        output_GMF = embed_user_GMF * embed_item_GMF

        embed_user_MLP = self.embed_user_MLP(user)
        embed_item_MLP = self.embed_item_MLP(item)
        interaction =  fluid.layers.concat(input=[embed_user_MLP, embed_item_MLP],axis=-1)
        output_MLP = self.MLP_layers(interaction)

        concat = fluid.layers.concat(input=[output_GMF, output_MLP], axis=-1)
        prediction = self.predict_layer(concat)
        return prediction



训练：


model = NCF(user_num=user_num, item_num=item_num,
                factor_num=factor_num, num_layers=num_layers, dropout=dropout)
    ##定义优化器
    adam = fluid.optimizer.AdamOptimizer(parameter_list=model.parameters())
   
   
    for epoch in range (epoch_num):
        model.train()
        for batch_id,data in enumerate(train_reader()):
            data =np.asarray(data).T
            user = fluid.dygraph.to_variable(data[0].reshape([-1,1]).astype("int64"))
            item = fluid.dygraph.to_variable(data[1].reshape([-1,1]).astype("int64"))
            label = fluid.dygraph.to_variable(data[2].reshape([-1,1]).astype("int64"))
            prediction = model(user,item)
            loss = fluid.layers.cross_entropy(input=prediction,label=label)
            # loss = fluid.layers.mse_loss(input=prediction,label=label)
            # loss = fluid.layers.square_error_cost(input=prediction,label=label)
            avg_loss = fluid.layers.mean(loss)
            # print(avg_loss.numpy())
            loss.backward()
            adam.minimize(avg_loss)
            model.clear_gradients()



> 你是不是设置了soft_label，方便提供下报错代码吗

**Comment by LDOUBLEV** [comment]
cross_entropy，对predict的最后一维是1时会出现这个错误，建议将网络的输出最后一维设置为2。正标签作为1类，负标签作为1类。

import paddle.fluid as fluid
import paddle.fluid.dygraph as dygraph
import numpy as np


with dygraph.guard():
    a = np.random.uniform(0, 1, [4, 2]).astype(np.float32)
    # b = np.array([[0], [1], [0], [1]]).astype(np.int64)
    b = np.random.randint(0, 2, [4, 1]).astype(np.int64)
    print(b)

    ap = dygraph.to_variable(a)
    bp = dygraph.to_variable(b)

    res = fluid.layers.cross_entropy(ap, bp)
    print(res)
